api/src/exam-environment/utils/exam.ts

```diff
@@ -575,6 +580,85 @@ export function generateExam(exam: EnvExam): Omit<EnvGeneratedExam, 'id'> {
   };
 }
 
+/**
+ * Calculates the number of correct questions over the number of the total questions given for an attempt.
+ * @returns The score of the exam attempt as a percentage.
+ */
+export function calculateScore(
+  exam: EnvExam,
+  generatedExam: EnvGeneratedExam,
+  attempt: EnvExamAttempt
+) {
+  const attemptQuestionSets = attempt.questionSets;
+  const generatedQuestionSets = generatedExam.questionSets;
+
+  const totalQuestions = generatedQuestionSets.reduce(
+    (total, attemptQuestionSet) => total + attemptQuestionSet.questions.length,
+    0
+  );
+  let correctQuestions = 0;
+  for (const attemptQuestionSet of attemptQuestionSets) {
+    const examQuestionSet = exam.questionSets.find(
+      ({ id }) => id === attemptQuestionSet.id
+    );
+    if (!examQuestionSet) {
+      throw new Error(
+        `Attempt question set ${attemptQuestionSet.id} must exist in exam ${exam.id}`
+      );
+    }
+
+    const generatedQuestionSet = generatedQuestionSets.find(
+      ({ id }) => id === attemptQuestionSet.id
+    );
+    if (!generatedQuestionSet) {
+      throw new Error(
+        `Generated question set ${attemptQuestionSet.id} must exist in generated exam ${generatedExam.id}`
+      );
+    }
+
+    const attemptQuestions = attemptQuestionSet.questions;
+    const examQuestions = examQuestionSet.questions;
+    const generatedQuestions = generatedQuestionSet.questions;
+    for (const attemptQuestion of attemptQuestions) {
+      const examQuestion = examQuestions.find(
+        ({ id }) => id === attemptQuestion.id
+      );
+      if (!examQuestion) {
+        throw new Error(
+          `Attempt question ${attemptQuestion.id} must exist in exam ${exam.id}`
+        );
+      }
+
+      const generatedQuestion = generatedQuestions.find(
+        ({ id }) => id === attemptQuestion.id
+      );
+      if (!generatedQuestion) {
+        throw new Error(
+          `Generated question ${attemptQuestion.id} must exist in generated exam ${generatedExam.id}`
+        );
+      }
+
+      // NOTE: The answers of an attempt is an array for future-proofing when
+      //       checkbox questions are needed.
+      //       This calculation takes x / y , x < y as wholey incorrect.
```

Comment:
What is this comment referring to?

Reply:
Which one?
- Part of it is explaining that it looks a bit stupid now to have an array of a single item, but this could make sense in the future
- The other part is explaining the marking as: "for `x` divided by `y`, where `x` is less than `y`, the question is marked as incorrect"
  - no half marks
  - each question is counted for as "one out of <total_num_questions>", as opposed to each correct answer counting for "one out of <total_num_correct_answers>"
  - `x`: correct answers correctly selected
  - `y`: total number of correct answers in question

Reply:
From the code, I get that a question is considered to be correct if all the answers related to that question are provided by the exam attempt and that the score is simply the percentage of the questions that are correct.

Unless I'm missing something, wouldn't that algorithm accept an attempt which marked ALL the checkboxes even if some of them were incorrect?

Reply:
Maybe... 🤫